runtime/cudaq/dynamics/callback.h

```diff
@@ -0,0 +1,112 @@
+#include <vector>
+
+namespace cudaq {
+
+class ScalarCallbackFunction {
+private:
+  // The user provided callback function that takes a map of complex 
+  // parameters.
+  std::function<std::complex<double>(const std::unordered_map<std::string, std::complex<double>>&)> _callback_func;
+
+public:
+  template <typename Callable>
+  ScalarCallbackFunction(Callable &&callable) {
+    static_assert(
+        std::is_invocable_r_v<std::complex<double>, Callable,
+                              const std::unordered_map<std::string, std::complex<double>>&>,
+        "Invalid callback function. Must have signature std::complex<double>("
+        "const std::unordered_map<std::string, std::complex<double>>&)");
+    _callback_func = std::forward<Callable>(callable);
+  }
+
+  // copy constructor
+  ScalarCallbackFunction(const ScalarCallbackFunction &other);
+
+  // move constructor.
+  ScalarCallbackFunction(ScalarCallbackFunction &&other);
+
+  // assignment operator
+  ScalarCallbackFunction& operator=(const ScalarCallbackFunction &other);
+
+  // move assignment operator
+  ScalarCallbackFunction& operator=(ScalarCallbackFunction &&other);
+
+  std::complex<double>
+  operator()(const std::unordered_map<std::string, std::complex<double>> &parameters) const;
+};
+
+
+class MatrixCallbackFunction {
+private:
+  // The user provided callback function that takes a vector defining the 
+  // dimension for each degree of freedom it acts on, and a map of complex 
+  // parameters.
+  std::function<matrix_2(const std::vector<int>&, const std::unordered_map<std::string, std::complex<double>>&)> _callback_func;
+
+public:
+  template <typename Callable>
+  MatrixCallbackFunction(Callable &&callable) {
+    static_assert(
+        std::is_invocable_r_v<matrix_2, Callable, const std::vector<int>&,
+                              const std::unordered_map<std::string, std::complex<double>>&>,
+        "Invalid callback function. Must have signature "
+        "matrix_2(const std::vector<int>&, const std::unordered_map<std::string, std::complex<double>>&)");
+    _callback_func = std::forward<Callable>(callable);
+  }
+
+  // copy constructor
+  MatrixCallbackFunction(const MatrixCallbackFunction &other);
+
+  // move constructor.
+  MatrixCallbackFunction(MatrixCallbackFunction &&other);
+
+  // assignment operator
+  MatrixCallbackFunction& operator=(const MatrixCallbackFunction &other);
+
+  // move assignment operator
+  MatrixCallbackFunction& operator=(MatrixCallbackFunction &&other);
+
+  matrix_2
+  operator()(const std::vector<int> &relevant_dimensions,
+             const std::unordered_map<std::string, std::complex<double>> &parameters) const;
+};
+
+
+/// @brief Object used to store the definition of a custom matrix operator.
+class Definition {
+private: 
+  std::string id;
+  MatrixCallbackFunction generator;
+  std::vector<int> required_dimensions;
+
+public:
+  const std::vector<int>& expected_dimensions = this->required_dimensions;
```

Comment:
Q: Was it meant to be a member function, like `const std::vector<int>& expected_dimensions() const { return this->required_dimensions;}` or a member variable?

Reply:
What's wrong with the const & as a variable (as written)? The purpose was to make sure the required dimensions cannot be modified externally, but it is just a stored value. What's the best C++ way to do that?